include/nfl/core.hpp

```diff
@@ -322,6 +325,68 @@ void poly<T, Degree, NbModuli>::set(gaussian<in_class, T, _lu_depth> const& mode
 #endif
 }
 
+template<class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(ZO_dist const& mode) {
+  set(mode);
+}
+
+template<class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set(ZO_dist const& mode) {
+  assert(mode.rho >= 0. && mode.rho <= 1.0);
+  uint8_t rnd[Degree];
+  value_type *ptr = &_data[0];
+  fastrandombytes(rnd, sizeof(rnd));
+  for (uint8_t r : rnd)
+  {
+    double odd = r * 1. / 0xFF;
+    *ptr++ = odd > mode.rho ? (r & 2) - 1U : 0U;
+  }
+  const value_type *end = &_data[N];
+  while (ptr != end)
+  {
+    std::memcpy(ptr, ptr + degree, degree * sizeof(value_type));
+    ptr += degree;
+  }
+}
+
+template<class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(hwt_dist const& mode) {
+  set(mode);
+}
+
+template<class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set(hwt_dist const& mode) {
+  assert(mode.hwt <= Degree);
+  std::vector<size_t> hitted(mode.hwt);
+  std::iota(hitted.begin(), hitted.end(), 0U); // select the first hwt positions.
+  std::vector<size_t> rnd(hitted.size());
+  auto rnd_end = rnd.end();
+  auto rnd_ptr = rnd_end;
+  for (size_t k = mode.hwt; k < degree; ++k) 
+  {
+    if (rnd_ptr == rnd_end)
+    {
+      fastrandombytes((unsigned char *)rnd.data(), rnd.size() * sizeof(size_t));
+      rnd_ptr = rnd.begin();
+    }
+    size_t pos = (*rnd_ptr++) % k;
```

Comment:
Why would we reduce mod k for increasing k (which starts at mode.hwt ?

Reply:
This sample is not uniform in [0...k] as size_t is not necessarily a multiple of k.

The statistical distance of the resulting distribution w.r.p. to the target distribution will be larger than 2**{-security} for common values of security.

To correct it I suggest using a rejection step before this step. For example using the fact that a size_t can go up to SIZE_MAX define `q = SIZE_MAX / k` then check if `*rnd_ptr > q*k`. If so restart the whole process (as k is very small w.r.t. SIZE_MAX this will almost never happen, but is needed from a security point of view).

Reply:
* Does `fastrandombytes` samples bytes uniformly ? 
   If it does, I think `rnd % k` should give uniform random in [0 .... k).

Reply:
Yes fastrandombytes samples bytes uniformly.

Suppose `k = 200`. If you sample an element `rnd`in `[0..255]` and do `rnd%200` then the values in `[0..55]` will have two preimages (one in `[0..55]` and one in `[200..255]`) and the values in `[56..199]` only one.

Thus the values in `[0..55]` will have twice the probability to appear w.r.t. the ones  in `[56..199]`.

Reply:
* Yes, I see the difference,
* However, for now, `rnd` is ranged from [0 ..., 2^64), (size_t is 2^64, right ?) ; given the fact that `k < PolyDegree` which is significantly smaller than 2^64. Thus, I think it would be fine to use `rnd % k`